neuralmonkey/dataset.py

```diff
@@ -395,3 +407,22 @@ def shuffle(self):
     def series_ids(self) -> Iterable[str]:
         return (list(self.series_paths_and_readers.keys()) +
                 list(self.preprocess_series.keys()))
+
+    def add_series(self, name: str, series: Iterable[Any]) -> None:
+        raise NotImplementedError(
+            "Lazy dataset does not support adding series.")
+
+
+def _preprocessed_datasets(dataset: Dataset, kwargs: Dict[str, Any]) -> None:
+    """Apply dataset-level preprocessing."""
+    keys = [key for key in kwargs.keys() if PREPROCESSED_SERIES.match(key)]
+
+    for key in keys:
+        name = PREPROCESSED_SERIES.match(key).group(1)
+        preprocessor = kwargs[key]
+
+        if isinstance(dataset, Dataset):
+            new_series = list(preprocessor(dataset))
+            dataset.add_series(name, new_series)
+        elif isinstance(dataset, LazyDataset):
+            dataset.preprocess_series[name] = (None, preprocessor)
```

Comment:
nerozumim moc tomu add_series - proč to lazy dataset neumí.. Je mi jasný, že proto, že tam voláš list a tady zavolat nejde, ale nerozumim tý motivaci to udělat zrovna takhle